Index: contrib/bind9/lib/dns/tkey.c
===================================================================
--- contrib/bind9/lib/dns/tkey.c	(revision 285922)
+++ contrib/bind9/lib/dns/tkey.c	(working copy)
@@ -650,6 +650,7 @@ dns_tkey_processquery(dns_message_t *msg, dns_tkey
 		 * Try the answer section, since that's where Win2000
 		 * puts it.
 		 */
+		name = NULL;
 		if (dns_message_findname(msg, DNS_SECTION_ANSWER, qname,
 					 dns_rdatatype_tkey, 0, &name,
 					 &tkeyset) != ISC_R_SUCCESS) {
