Index: sys/ufs/ffs/ffs_inode.c
===================================================================
RCS file: /home/ncvs/src/sys/ufs/ffs/ffs_inode.c,v
retrieving revision 1.56.2.5
diff -u -r1.56.2.5 ffs_inode.c
--- sys/ufs/ffs/ffs_inode.c	5 Feb 2002 18:35:03 -0000	1.56.2.5
+++ sys/ufs/ffs/ffs_inode.c	11 Mar 2005 14:29:19 -0000
@@ -197,6 +197,7 @@
 #endif
 			softdep_setup_freeblocks(oip, length);
 			vinvalbuf(ovp, 0, cred, p, 0, 0);
+			vnode_pager_setsize(vp, 0);
 			oip->i_flag |= IN_CHANGE | IN_UPDATE;
 			return (ffs_update(ovp, 0));
 		}
