Index: crypto/openssl/ssl/s3_pkt.c
===================================================================
RCS file: /home/ncvs/src/crypto/openssl/ssl/s3_pkt.c,v
retrieving revision 1.1.1.8
diff -c -r1.1.1.8 s3_pkt.c
*** crypto/openssl/ssl/s3_pkt.c	19 Feb 2003 23:17:05 -0000	1.1.1.8
--- crypto/openssl/ssl/s3_pkt.c	16 Mar 2004 13:18:28 -0000
***************
*** 1085,1090 ****
--- 1085,1098 ----
  			goto err;
  			}
  
+ 		/* Check we have a cipher to change to */
+ 		if (s->s3->tmp.new_cipher == NULL)
+ 			{
+ 			i=SSL_AD_UNEXPECTED_MESSAGE;
+ 			SSLerr(SSL_F_SSL3_GET_CERT_VERIFY,SSL_R_CCS_RECEIVED_EARLY);
+ 			goto err;
+ 			}
+ 
  		rr->length=0;
  
  		if (s->msg_callback)
