Index: sys/dev/random/yarrow.c
===================================================================
RCS file: /usr/ncvs/src/sys/dev/random/yarrow.c,v
retrieving revision 1.47
diff -u -r1.47 yarrow.c
--- sys/dev/random/yarrow.c	27 May 2007 18:54:58 -0000	1.47
+++ sys/dev/random/yarrow.c	27 Nov 2007 17:17:29 -0000
@@ -296,6 +296,7 @@
 				random_state.outputblocks = 0;
 			}
 			retval += (int)tomove;
+			cur = 0;
 		}
 	}
 	else {
