--- usr.bin/quota/quota.c.orig
+++ usr.bin/quota/quota.c
@@ -606,7 +606,7 @@
 	call_stat = callaurpc(host, RQUOTAPROG, EXT_RQUOTAVERS,
 			      RQUOTAPROC_GETQUOTA, (xdrproc_t)xdr_ext_getquota_args, (char *)&gq_args,
 			      (xdrproc_t)xdr_getquota_rslt, (char *)&gq_rslt);
-	if (call_stat == RPC_PROGVERSMISMATCH) {
+	if (call_stat == RPC_PROGVERSMISMATCH || call_stat == RPC_PROGNOTREGISTERED) {
 		if (quotatype == USRQUOTA) {
 			old_gq_args.gqa_pathp = cp + 1;
 			old_gq_args.gqa_uid = id;
