Index: contrib/bind9/bin/named/update.c
===================================================================
--- contrib/bind9/bin/named/update.c	(revision 195863)
+++ contrib/bind9/bin/named/update.c	(working copy)
@@ -979,7 +979,11 @@
 			if (type == dns_rdatatype_rrsig ||
 			    type == dns_rdatatype_sig)
 				covers = dns_rdata_covers(&t->rdata);
-			else
+			else if (type == dns_rdatatype_any) {
+				dns_db_detachnode(db, &node);
+				dns_diff_clear(&trash);
+				return (DNS_R_NXRRSET);
+			} else
 				covers = 0;
 
 			/*
