Index: sys/dev/bce/if_bce.c
===================================================================
RCS file: /home/ncvs/src/sys/dev/bce/if_bce.c,v
retrieving revision 1.34.2.7
retrieving revision 1.34.2.8
diff -p -I __FBSDID -I $FreeBSD -u -u -r1.34.2.7 -r1.34.2.8
--- sys/dev/bce/if_bce.c	31 Mar 2009 01:01:01 -0000	1.34.2.7
+++ sys/dev/bce/if_bce.c	20 May 2009 21:13:49 -0000	1.34.2.8
@@ -5895,6 +5895,9 @@ bce_rx_intr(struct bce_softc *sc)
 			/* Set the total packet length. */
 			m0->m_pkthdr.len = m0->m_len = pkt_len;
 		}
+#else
+        /* Set the total packet length. */
+		m0->m_pkthdr.len = m0->m_len = pkt_len;
 #endif
 
 		/* Remove the trailing Ethernet FCS. */
