Index: sys/dev/iir/iir_ctrl.c
===================================================================
RCS file: /home/ncvs/src/sys/dev/iir/iir_ctrl.c,v
retrieving revision 1.16
diff -u -p -r1.16 iir_ctrl.c
--- sys/dev/iir/iir_ctrl.c	6 Jan 2005 01:42:47 -0000	1.16
+++ sys/dev/iir/iir_ctrl.c	3 May 2005 11:44:40 -0000
@@ -102,12 +102,12 @@ gdt_make_dev(int unit)
 
 #ifdef SDEV_PER_HBA
     dev = make_dev(&iir_cdevsw, hba2minor(unit), UID_ROOT, GID_OPERATOR,
-                   S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH, "iir%d", unit);
+                   S_IRUSR | S_IWUSR, "iir%d", unit);
 #else
     if (sdev_made)
         return (0);
     dev = make_dev(&iir_cdevsw, 0, UID_ROOT, GID_OPERATOR,
-                   S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH, "iir");
+                   S_IRUSR | S_IWUSR, "iir");
     sdev_made = 1;
 #endif
     return (dev);
