Index: sys/netinet6/nd6.c
===================================================================
RCS file: /home/ncvs/src/sys/netinet6/nd6.c,v
retrieving revision 1.48.2.15
diff -u -r1.48.2.15 nd6.c
--- sys/netinet6/nd6.c	7 Oct 2006 18:31:27 -0000	1.48.2.15
+++ sys/netinet6/nd6.c	15 Feb 2007 02:34:00 -0000
@@ -1315,7 +1315,7 @@
 		callout_init(&ln->ln_timer_ch, 0);
 
 		/* this is required for "ndp" command. - shin */
-		if (req == RTM_ADD && (rt->rt_flags & RTF_STATIC)) {
+		if (req == RTM_ADD) {
 		        /*
 			 * gate should have some valid AF_LINK entry,
 			 * and ln->ln_expire should have some lifetime
