Index: usr.bin/fetch/fetch.c
===================================================================
RCS file: /home/ncvs/src/usr.bin/fetch/fetch.c,v
retrieving revision 1.74
diff -u -p -r1.74 fetch.c
--- usr.bin/fetch/fetch.c	21 Sep 2004 18:34:19 -0000	1.74
+++ usr.bin/fetch/fetch.c	14 Nov 2004 14:03:12 -0000
@@ -584,7 +584,8 @@ fetch(char *URL, const char *path)
 	/* suck in the data */
 	signal(SIGINFO, sig_handler);
 	while (!sigint) {
-		if (us.size != -1 && us.size - count < B_size)
+		if (us.size != -1 && us.size - count < B_size &&
+		    us.size - count >= 0)
 			size = us.size - count;
 		else
 			size = B_size;
