Index: secure/lib/libcrypt/crypt-des.c
===================================================================
--- secure/lib/libcrypt/crypt-des.c	(revision 234682)
+++ secure/lib/libcrypt/crypt-des.c	(working copy)
@@ -606,7 +606,7 @@
 	q = (u_char *)keybuf;
 	while (q - (u_char *)keybuf - 8) {
 		*q++ = *key << 1;
-		if (*(q - 1))
+		if (*key != '\0')
 			key++;
 	}
 	if (des_setkey((char *)keybuf))
