Index: usr.sbin/pw/pw.c
===================================================================
--- usr.sbin/pw/pw.c	(revision 287410)
+++ usr.sbin/pw/pw.c	(working copy)
@@ -272,14 +272,7 @@
 				    errstr);
 			break;
 		case 'n':
-			if (strspn(optarg, "0123456789") != strlen(optarg)) {
-				name = optarg;
-				break;
-			}
-			id = strtonum(optarg, 0, LONG_MAX, &errstr);
-			if (errstr != NULL)
-				errx(EX_USAGE, "Bad id '%s': %s", optarg,
-				    errstr);
+			name = optarg;
 			break;
 		case 'o':
 			conf.checkduplicate = false;
