--- usr.sbin/freebsd-update/freebsd-update.sh.orig
+++ usr.sbin/freebsd-update/freebsd-update.sh
@@ -2576,7 +2576,7 @@
 
 			# Some files need special treatment.
 			case ${F} in
-			/etc/spwd.db | /etc/pwd.db | /etc/login.conf.db)
+			/etc/spwd.db | /etc/pwd.db | /etc/login.conf.db | /etc/ssl/cert.pem)
 				# Don't merge these -- we're rebuild them
 				# after updates are installed.
 				cp merge/old/${F} merge/new/${F}
