Index: sys/dev/firewire/fwdev.c
===================================================================
RCS file: /home/ncvs/src/sys/dev/firewire/fwdev.c,v
retrieving revision 1.47
diff -u -I__FBSDID -r1.47 fwdev.c
--- sys/dev/firewire/fwdev.c	8 Aug 2005 19:55:30 -0000	1.47
+++ sys/dev/firewire/fwdev.c	30 Nov 2006 22:15:06 -0000
@@ -712,7 +712,7 @@
 			else
 				len = fwdev->rommax - CSRROMOFF + 4;
 		}
-		if (crom_buf->len < len)
+		if (crom_buf->len < len && crom_buf->len >= 0)
 			len = crom_buf->len;
 		else
 			crom_buf->len = len;
