--- sys/net/rtsock.c	2003/12/08 17:16:58	1.44.2.12
+++ sys/net/rtsock.c	2004/04/06 20:11:53	1.44.2.13
@@ -330,8 +330,8 @@
 	 * Verify that the caller has the appropriate privilege; RTM_GET
 	 * is the only operation the non-superuser is allowed.
 	 */
-	if (rtm->rtm_type != RTM_GET && suser_xxx(so->so_cred, NULL, 0) != 0)
-		senderr(EPERM);
+	if (rtm->rtm_type != RTM_GET && (error = suser(curproc)) != 0)
+		senderr(error);
 
 	switch (rtm->rtm_type) {
 
