Index: sys/kern/kern_descrip.c
diff -c sys/kern/kern_descrip.c:1.81.2.14 src/sys/kern/kern_descrip.c:1.81.2.15
*** sys/kern/kern_descrip.c:1.81.2.14	Mon Apr 29 10:14:12 2002
--- sys/kern/kern_descrip.c	Sun Nov 10 19:43:31 2002
***************
*** 711,720 ****
  	switch (fp->f_type) {
  	case DTYPE_PIPE:
  	case DTYPE_SOCKET:
! 		if (uap->name != _PC_PIPE_BUF)
! 			return (EINVAL);
! 		p->p_retval[0] = PIPE_BUF;
! 		error = 0;
  		break;
  	case DTYPE_FIFO:
  	case DTYPE_VNODE:
--- 711,722 ----
  	switch (fp->f_type) {
  	case DTYPE_PIPE:
  	case DTYPE_SOCKET:
! 		if (uap->name != _PC_PIPE_BUF) {
! 			error = EINVAL;
! 		} else {
! 			p->p_retval[0] = PIPE_BUF;
! 			error = 0;
! 		}
  		break;
  	case DTYPE_FIFO:
  	case DTYPE_VNODE:
