Index: contrib/cvs/src/cvsbug.in
===================================================================
RCS file: /home/ncvs/src/contrib/cvs/src/cvsbug.in,v
retrieving revision 1.1.1.1.2.2
diff -u -p -r1.1.1.1.2.2 cvsbug.in
--- contrib/cvs/src/cvsbug.in	19 Dec 2002 21:17:56 -0000	1.1.1.1.2.2
+++ contrib/cvs/src/cvsbug.in	9 Sep 2005 06:52:35 -0000
@@ -85,9 +85,9 @@ fi
 
 [ -z "$TMPDIR" ] && TMPDIR=/tmp
 
-TEMP=$TMPDIR/p$$
-BAD=$TMPDIR/pbad$$
-REF=$TMPDIR/pf$$
+TEMP="`/usr/bin/mktemp $TMPDIR/p.XXXXXX`"
+BAD="`/usr/bin/mktemp $TMPDIR/pbad.XXXXXX`"
+REF="`/usr/bin/mktemp $TMPDIR/pf.XXXXXX`"
 
 if [ -z "$LOGNAME" -a -n "$USER" ]; then
   LOGNAME=$USER
@@ -108,14 +108,12 @@ elif [ -f /bin/domainname ]; then
     /usr/bin/ypcat passwd 2>/dev/null | cat - /etc/passwd | grep "^$LOGNAME:" |
       cut -f5 -d':' | sed -e 's/,.*//' > $TEMP
     ORIGINATOR="`cat $TEMP`"
-    rm -f $TEMP
   fi
 fi
 
 if [ "$ORIGINATOR" = "" ]; then
   grep "^$LOGNAME:" /etc/passwd | cut -f5 -d':' | sed -e 's/,.*//' > $TEMP
   ORIGINATOR="`cat $TEMP`"
-  rm -f $TEMP
 fi
 
 if [ -n "$ORGANIZATION" ]; then
Index: gnu/usr.bin/send-pr/send-pr.sh
===================================================================
RCS file: /home/ncvs/src/gnu/usr.bin/send-pr/send-pr.sh,v
retrieving revision 1.13.2.13
diff -u -p -r1.13.2.13 send-pr.sh
--- gnu/usr.bin/send-pr/send-pr.sh	1 Dec 2003 19:12:58 -0000	1.13.2.13
+++ gnu/usr.bin/send-pr/send-pr.sh	9 Sep 2005 06:52:35 -0000
@@ -262,7 +262,7 @@ TEMP=`mktemp -t pf` || exit 1
 # Catch some signals. ($xs kludge needed by Sun /bin/sh)
 xs=0
 trap 'rm -f $REF $TEMP; exit $xs' 0
-trap 'echo "$COMMAND: Aborting ... saving unfinished PR into /tmp/pr.$$"; rm -f $REF ; mv $TEMP /tmp/pr.$$; xs=1; exit' 1 2 3 13 15
+trap 'SAV=`mktemp -t pr`;echo "$COMMAND: Aborting ... saving unfinished PR into $SAV"; rm -f $REF ; mv $TEMP $SAV; xs=1; exit' 1 2 3 13 15
 
 # If they told us to use a specific file, then do so.
 if [ -n "$IN_FILE" ]; then
