Index: contrib/cvs/src/cvsbug.in
===================================================================
RCS file: /home/ncvs/src/contrib/cvs/src/cvsbug.in,v
retrieving revision 1.1.1.3
diff -u -p -r1.1.1.3 cvsbug.in
--- contrib/cvs/src/cvsbug.in	15 Apr 2004 01:01:55 -0000	1.1.1.3
+++ contrib/cvs/src/cvsbug.in	5 Sep 2005 06:23:21 -0000
@@ -109,14 +109,12 @@ elif [ -f /bin/domainname ]; then
     /usr/bin/ypcat passwd 2>/dev/null | cat - /etc/passwd | grep "^$LOGNAME:" |
       cut -f5 -d':' | sed -e 's/,.*//' > $TEMP
     ORIGINATOR="`cat $TEMP`"
-    rm -f $TEMP
   fi
 fi
 
 if [ "$ORIGINATOR" = "" ]; then
   grep "^$LOGNAME:" /etc/passwd | cut -f5 -d':' | sed -e 's/,.*//' > $TEMP
   ORIGINATOR="`cat $TEMP`"
-  rm -f $TEMP
 fi
 
 if [ -n "$ORGANIZATION" ]; then
