--- crypto/openssh/sshd.c.orig
+++ crypto/openssh/sshd.c
@@ -377,8 +377,6 @@
 		kill(0, SIGTERM);
 	}
 
-	BLACKLIST_NOTIFY(the_active_state, BLACKLIST_AUTH_FAIL, "ssh");
-
 	/* Log error and exit. */
 	sigdie("Timeout before authentication for %s port %d",
 	    ssh_remote_ipaddr(the_active_state),
--- crypto/openssh/version.h.orig
+++ crypto/openssh/version.h
@@ -5,4 +5,4 @@
 #define SSH_PORTABLE	"p1"
 #define SSH_RELEASE	SSH_VERSION SSH_PORTABLE
 
-#define SSH_VERSION_FREEBSD	"FreeBSD-20240701"
+#define SSH_VERSION_FREEBSD	"FreeBSD-20240806"
