Index: contrib/ntp/ntpd/ntp_crypto.c
===================================================================
--- contrib/ntp/ntpd/ntp_crypto.c	(revision 186798)
+++ contrib/ntp/ntpd/ntp_crypto.c	(working copy)
@@ -1536,7 +1536,7 @@
 		EVP_VerifyUpdate(&ctx, (u_char *)&ep->tstamp, vallen +
 		    12);
 		if (EVP_VerifyFinal(&ctx, (u_char *)&ep->pkt[i], siglen,
-		    pkey)) {
+		    pkey) == 1) {
 			if (peer->crypto & CRYPTO_FLAG_VRFY)
 				peer->crypto |= CRYPTO_FLAG_PROV;
 		} else {
