Index: contrib/bind9/lib/dns/message.c
===================================================================
--- contrib/bind9/lib/dns/message.c	(revision 223383)
+++ contrib/bind9/lib/dns/message.c	(working copy)
@@ -2481,7 +2481,7 @@
 	    msg->opcode != dns_opcode_notify)
 		want_question_section = ISC_FALSE;
 	if (msg->opcode == dns_opcode_update)
-		first_section = DNS_SECTION_ADDITIONAL;
+		first_section = DNS_SECTION_PREREQUISITE;
 	else if (want_question_section) {
 		if (!msg->question_ok)
 			return (DNS_R_FORMERR);
