Index: sys/net/if_bridge.c
===================================================================
RCS file: /usr/ncvs/src/sys/net/if_bridge.c,v
retrieving revision 1.50
diff -u -r1.50 if_bridge.c
--- sys/net/if_bridge.c	14 Jan 2006 03:51:30 -0000	1.50
+++ sys/net/if_bridge.c	22 Jan 2006 18:05:34 -0000
@@ -615,6 +615,7 @@
 			break;
 		}
 
+		bzero(&args, sizeof args);
 		if (bc->bc_flags & BC_F_COPYIN) {
 			error = copyin(ifd->ifd_data, &args, ifd->ifd_len);
 			if (error)
@@ -1022,6 +1023,7 @@
 
 	count = 0;
 	len = bifc->ifbic_len;
+	bzero(&breq, sizeof breq);
 	LIST_FOREACH(bif, &sc->sc_iflist, bif_next) {
 		if (len < sizeof(breq))
 			break;
@@ -1075,6 +1077,7 @@
 		return (0);
 
 	len = bac->ifbac_len;
+	bzero(&bareq, sizeof bareq);
 	LIST_FOREACH(brt, &sc->sc_rtlist, brt_list) {
 		if (len < sizeof(bareq))
 			goto out;
