Index: etc/rc
diff -u etc/rc:1.212.2.49 etc/rc:1.212.2.50
--- etc/rc:1.212.2.49	Sun Apr 28 17:49:59 2002
+++ etc/rc	Thu May  9 12:39:01 2002
@@ -370,7 +370,9 @@
 # Remove X lock files, since they will prevent you from restarting X11
 # after a system crash.
 #
-rm -f /tmp/.X*-lock /tmp/.X11-unix/*
+rm -f /tmp/.X*-lock
+rm -fr /tmp/.X11-unix
+mkdir -m 1777 /tmp/.X11-unix
 
 # Snapshot any kernel -c changes back to disk here <someday>.
 # This has changed with ELF and /kernel.config.
