Index: crypto/heimdal/lib/gssapi/krb5/prf.c
===================================================================
--- crypto/heimdal/lib/gssapi/krb5/prf.c	(revision 267806)
+++ crypto/heimdal/lib/gssapi/krb5/prf.c	(working copy)
@@ -117,7 +117,7 @@ _gsskrb5_pseudo_random(OM_uint32 *minor_status,
     num = 0;
     p = prf_out->value;
     while(desired_output_len > 0) {
-	_gsskrb5_encode_om_uint32(num, input.data);
+	_gsskrb5_encode_be_om_uint32(num, input.data);
 	ret = krb5_crypto_prf(context, crypto, &input, &output);
 	if (ret) {
 	    OM_uint32 junk;
