Index: sshd.c
===================================================================
RCS file: /home/ncvs/src/crypto/openssh/sshd.c,v
retrieving revision 1.6
diff -u -r1.6 sshd.c
--- sshd.c	2000/03/09 14:52:31	1.6
+++ sshd.c	2000/07/04 09:27:02
@@ -2564,7 +2564,13 @@
 	char *argv[10];
 #ifdef LOGIN_CAP
 	login_cap_t *lc;
+#endif
 
+	/* login(1) is only called if we execute the login shell */
+	if (options.use_login && command != NULL)
+		options.use_login = 0;
+
+#ifdef LOGIN_CAP
 	lc = login_getpwclass(pw);
 	if (lc == NULL)
 		lc = login_getclassbyname(NULL, pw);
