Index: builtins.c
===================================================================
RCS file: /home/ncvs/src/usr.sbin/inetd/builtins.c,v
retrieving revision 1.14.2.1
diff -u -r1.14.2.1 builtins.c
--- builtins.c	1999/11/18 09:33:47	1.14.2.1
+++ builtins.c	2001/01/03 23:47:00
@@ -473,8 +473,9 @@
 		 * open any files we have no permission to open, especially
 		 * symbolic links to sensitive root-owned files or devices.
 		 */
+		if (initgroups(pw->pw_name, pw->pw_gid) == -1)
+			iderror(lport, fport, s, errno);
 		seteuid(pw->pw_uid);
-		setegid(pw->pw_gid);
 		/*
 		 * If we were to lstat() here, it would do no good, since it
 		 * would introduce a race condition and could be defeated.
