Index: lib/bind/config.h
===================================================================
--- lib/bind/config.h	(revision 248771)
+++ lib/bind/config.h	(working copy)
@@ -281,7 +281,7 @@ int sigwait(const unsigned int *set, int *sig);
 /* #undef HAVE_OPENSSL_GOST */
 
 /* Define to 1 if you have the <regex.h> header file. */
-#define HAVE_REGEX_H 1
+/* #undef HAVE_REGEX_H */
 
 /* Define to 1 if you have the `setegid' function. */
 #define HAVE_SETEGID 1
