Index: usr.sbin/freebsd-update/freebsd-update.sh
===================================================================
--- usr.sbin/freebsd-update/freebsd-update.sh	(revision 199972)
+++ usr.sbin/freebsd-update/freebsd-update.sh	(working copy)
@@ -603,6 +603,7 @@
 		echo ${WORKDIR}
 		exit 1
 	fi
+	chmod 700 ${WORKDIR}
 	cd ${WORKDIR} || exit 1
 
 	# Generate release number.  The s/SECURITY/RELEASE/ bit exists
Index: etc/mtree/BSD.var.dist
===================================================================
--- etc/mtree/BSD.var.dist	(revision 200019)
+++ etc/mtree/BSD.var.dist	(working copy)
@@ -32,7 +32,7 @@
     db
         entropy         uname=operator gname=operator mode=0700
         ..
-        freebsd-update
+        freebsd-update  mode=0700
         ..
         ipf             mode=0700
         ..
