Index: sys/i386/i386/initcpu.c
===================================================================
--- sys/i386/i386/initcpu.c	(revision 347631)
+++ sys/i386/i386/initcpu.c	(revision 347632)
@@ -745,7 +745,6 @@
 		cpu_fxsr = hw_instruction_sse = 1;
 	}
 #if defined(PAE) || defined(PAE_TABLES)
-	hw_mds_recalculate();
 	if ((amd_feature & AMDID_NX) != 0) {
 		uint64_t msr;
 
@@ -755,6 +754,7 @@
 		elf32_nxstack = 1;
 	}
 #endif
+	hw_mds_recalculate();
 }
 
 void
