Index: contrib/sendmail/src/trace.c
===================================================================
RCS file: /home/ncvs/src/contrib/sendmail/src/trace.c,v
retrieving revision 1.1.1.2.6.2
retrieving revision 1.1.1.2.6.2.2.1
diff -u -r1.1.1.2.6.2 -r1.1.1.2.6.2.2.1
--- contrib/sendmail/src/trace.c	2000/10/10 05:07:28	1.1.1.2.6.2
+++ contrib/sendmail/src/trace.c	2001/08/22 05:34:11	1.1.1.2.6.2.2.1
@@ -63,7 +63,7 @@
 tTflag(s)
 	register char *s;
 {
-	int first, last;
+	unsigned int first, last;
 	register unsigned int i;
 
 	if (*s == '\0')
