--- sys/cam/cam_periph.c.orig
+++ sys/cam/cam_periph.c
@@ -52,6 +52,7 @@
 
 #include <cam/cam.h>
 #include <cam/cam_ccb.h>
+#include <cam/cam_compat.h>
 #include <cam/cam_queue.h>
 #include <cam/cam_xpt_periph.h>
 #include <cam/cam_xpt_internal.h>
@@ -1120,6 +1121,7 @@
 	error = found = 0;
 
 	switch(cmd){
+	case CAMGETPASSTHRU_0x19:
 	case CAMGETPASSTHRU:
 		ccb = cam_periph_getccb(periph, CAM_PRIORITY_NORMAL);
 		xpt_setup_ccb(&ccb->ccb_h,
