Index: contrib/bind9/lib/dns/ncache.c
===================================================================
--- contrib/bind9/lib/dns/ncache.c	(revision 220971)
+++ contrib/bind9/lib/dns/ncache.c	(working copy)
@@ -186,7 +186,7 @@
 					 */
 					isc_buffer_availableregion(&buffer,
 								   &r);
-					if (r.length < 2)
+					if (r.length < 3)
 						return (ISC_R_NOSPACE);
 					isc_buffer_putuint16(&buffer,
 							     rdataset->type);
