Index: contrib/bind9/lib/dns/validator.c
===================================================================
RCS file: /home/ncvs/src/contrib/bind9/lib/dns/validator.c,v
retrieving revision 1.1.1.1.2.1
diff -u -p -r1.1.1.1.2.1 validator.c
--- contrib/bind9/lib/dns/validator.c	26 Sep 2004 03:09:51 -0000	1.1.1.1.2.1
+++ contrib/bind9/lib/dns/validator.c	8 Jun 2005 20:29:12 -0000
@@ -497,6 +497,8 @@
 
 	REQUIRE(exists != NULL);
 	REQUIRE(data != NULL);
+	REQUIRE(nsecset != NULL &&
+	nsecset->type == dns_rdatatype_nsec);
 
 	result = dns_rdataset_first(nsecset);
 	if (result != ISC_R_SUCCESS) {
@@ -661,7 +663,7 @@
 		if (rdataset->trust == dns_trust_secure)
 			val->seensig = ISC_TRUE;
 
-		if (val->nsecset != NULL &&
+		if (rdataset->type == dns_rdatatype_nsec &&
 		    rdataset->trust == dns_trust_secure &&
 		    ((val->attributes & VALATTR_NEEDNODATA) != 0 ||
 		     (val->attributes & VALATTR_NEEDNOQNAME) != 0) &&
