Index: sys/contrib/pf/net/pf_norm.c
===================================================================
RCS file: /home/ncvs/src/sys/contrib/pf/net/pf_norm.c,v
retrieving revision 1.11.2.2
diff -u -p -I__FBSDID -r1.11.2.2 pf_norm.c
--- sys/contrib/pf/net/pf_norm.c	17 Jan 2006 13:05:32 -0000	1.11.2.2
+++ sys/contrib/pf/net/pf_norm.c	22 Jan 2006 16:38:31 -0000
@@ -818,7 +818,7 @@ pf_fragcache(struct mbuf **m0, struct ip
 			} else {
 				hosed++;
 			}
-		} else {
+		} else if (frp == NULL) {
 			/* There is a gap between fragments */
 			DPFPRINTF(("fragcache[%d]: gap %d %d-%d (%d-%d)\n",
 			    h->ip_id, -aftercut, off, max, fra->fr_off,
