Index: sys/netinet6/ipcomp_input.c
===================================================================
RCS file: /home/ncvs/src/sys/netinet6/Attic/ipcomp_input.c,v
retrieving revision 1.7.4.1
diff -u -r1.7.4.1 ipcomp_input.c
--- sys/netinet6/ipcomp_input.c	31 Jan 2005 23:26:39 -0000	1.7.4.1
+++ sys/netinet6/ipcomp_input.c	13 Feb 2008 13:44:24 -0000
@@ -258,7 +258,7 @@
 	off = *offp;
 
 	md = m_pulldown(m, off, sizeof(*ipcomp), NULL);
-	if (!m) {
+	if (!md) {
 		m = NULL;	/* already freed */
 		ipseclog((LOG_DEBUG, "IPv6 IPComp input: assumption failed "
 		    "(pulldown failure)\n"));
