Index: contrib/sendmail/src/parseaddr.c
===================================================================
RCS file: /home/ncvs/src/contrib/sendmail/src/parseaddr.c,v
retrieving revision 1.1.1.2.6.13
retrieving revision 1.1.1.2.6.14
diff -c -p -c -r1.1.1.2.6.13 -r1.1.1.2.6.14
*** contrib/sendmail/src/parseaddr.c	29 Mar 2003 19:33:17 -0000	1.1.1.2.6.13
--- contrib/sendmail/src/parseaddr.c	17 Sep 2003 15:18:20 -0000	1.1.1.2.6.14
*************** prescan(addr, delim, pvpbuf, pvpbsize, d
*** 700,706 ****
--- 700,710 ----
  						addr[MAXNAME] = '\0';
  	returnnull:
  					if (delimptr != NULL)
+ 					{
+ 						if (p > addr)
+ 							p--;
  						*delimptr = p;
+ 					}
  					CurEnv->e_to = saveto;
  					return NULL;
  				}
