Index: etc/rc.d/jail
===================================================================
RCS file: /home/ncvs/src/etc/rc.d/jail,v
retrieving revision 1.15.2.5.2.1
diff -u -d -r1.15.2.5.2.1 jail
--- etc/rc.d/jail	11 Jan 2007 18:19:33 -0000	1.15.2.5.2.1
+++ etc/rc.d/jail	27 Jul 2007 08:49:37 -0000
@@ -228,7 +228,7 @@
 			warn "${_mountpt} has symlink as parent - not mounting from ${jail_fstab}"
 			return
 		fi
-	done <${_fstab}
+	done <${jail_fstab}
 	mount -a -F "${jail_fstab}"
 }
 
