Index: sys/kern/kern_descrip.c
===================================================================
RCS file: /home/ncvs/src/sys/kern/kern_descrip.c,v
retrieving revision 1.313
diff -u -d -r1.313 kern_descrip.c
--- sys/kern/kern_descrip.c	6 Aug 2007 14:26:00 -0000	1.313
+++ sys/kern/kern_descrip.c	4 Feb 2008 19:05:42 -0000
@@ -2046,7 +2046,7 @@
 	int error;
 
 	*vpp = NULL;
-	if ((error = _fget(td, fd, &fp, 0, 0)) != 0)
+	if ((error = _fget(td, fd, &fp, flags, 0)) != 0)
 		return (error);
 	if (fp->f_vnode == NULL) {
 		error = EINVAL;
