Index: sys/kern/vfs_mount.c
===================================================================
RCS file: /usr/ncvs/src/sys/kern/vfs_mount.c,v
retrieving revision 1.265.2.3
diff -u -r1.265.2.3 vfs_mount.c
--- sys/kern/vfs_mount.c	6 Apr 2008 10:02:20 -0000	1.265.2.3
+++ sys/kern/vfs_mount.c	17 Jul 2008 15:39:37 -0000
@@ -1830,7 +1830,8 @@
 		}
 		if (*t != NULL)
 			continue;
-		sprintf(errmsg, "mount option <%s> is unknown", p);
+		snprintf(errmsg, sizeof(errmsg),
+		    "mount option <%s> is unknown", p);
 		printf("%s\n", errmsg);
 		ret = EINVAL;
 	}
