Index: sys/netinet/igmp.c
===================================================================
--- sys/netinet/igmp.c	(revision 278973)
+++ sys/netinet/igmp.c	(working copy)
@@ -1533,8 +1533,8 @@ igmp_input(struct mbuf *m, int off)
 		case IGMP_VERSION_3: {
 				struct igmpv3 *igmpv3;
 				uint16_t igmpv3len;
-				uint16_t srclen;
-				int nsrc;
+				uint16_t nsrc;
+				int srclen;
 
 				IGMPSTAT_INC(igps_rcv_v3_queries);
 				igmpv3 = (struct igmpv3 *)igmp;
