Index: etc/defaults/rc.conf
===================================================================
--- etc/defaults/rc.conf	(revision 265059)
+++ etc/defaults/rc.conf	(working copy)
@@ -649,7 +649,7 @@
 devfs_system_ruleset=""	# The name (NOT number) of a ruleset to apply to /dev
 devfs_set_rulesets=""	# A list of /mount/dev=ruleset_name settings to
 			# apply (must be mounted already, i.e. fstab(5))
-devfs_load_rulesets="NO"	# Enable to always load the default rulesets
+devfs_load_rulesets="YES"	# Enable to always load the default rulesets
 performance_cx_lowest="HIGH"	# Online CPU idle state
 performance_cpu_freq="NONE"	# Online CPU frequency
 economy_cx_lowest="HIGH"	# Offline CPU idle state
