Index: contrib/bind9/lib/dns/zone.c
===================================================================
--- contrib/bind9/lib/dns/zone.c	(revision 278973)
+++ contrib/bind9/lib/dns/zone.c	(working copy)
@@ -8496,6 +8496,12 @@ keyfetch_done(isc_task_t *task, isc_event_t *event
 					     namebuf, tag);
 				trustkey = ISC_TRUE;
 			}
+		} else {
+			/*
+			 * No previously known key, and the key is not
+			 * secure, so skip it.
+			 */
+			continue;
 		}
 
 		/* Delete old version */
@@ -8544,7 +8550,7 @@ keyfetch_done(isc_task_t *task, isc_event_t *event
 			trust_key(zone, keyname, &dnskey, mctx);
 		}
 
-		if (!deletekey)
+		if (secure && !deletekey)
 			set_refreshkeytimer(zone, &keydata, now);
 	}
 
