--- contrib/tcp_wrappers/socket.c	2000/09/25 00:41:55	1.5
+++ contrib/tcp_wrappers/socket.c	2001/07/04 20:16:18	1.6
@@ -222,7 +222,7 @@
 	hints.ai_family = sin->sa_family;
 	hints.ai_socktype = SOCK_STREAM;
 	hints.ai_flags = AI_PASSIVE | AI_CANONNAME | AI_NUMERICHOST;
-	if ((err = getaddrinfo(host->name, NULL, &hints, &res0) == 0)) {
+	if ((err = getaddrinfo(host->name, NULL, &hints, &res0)) == 0) {
 	    freeaddrinfo(res0);
 	    tcpd_warn("host name/name mismatch: "
 		      "reverse lookup results in non-FQDN %s",
