Index: channels.c
===================================================================
RCS file: /home/ncvs/src/crypto/openssh/channels.c,v
retrieving revision 1.7
retrieving revision 1.8
diff -u -r1.7 -r1.8
--- crypto/openssh/channels.c	23 Jan 2002 15:06:47 -0000	1.7
+++ crypto/openssh/channels.c	5 Mar 2002 14:27:19 -0000	1.8
@@ -151,7 +151,7 @@
 channel_lookup(int id)
 {
 	Channel *c;
-	if (id < 0 || id > channels_alloc) {
+	if (id < 0 || id >= channels_alloc) {
 		log("channel_lookup: %d: bad id", id);
 		return NULL;
 	}
