Index: gnu/libexec/uucp/cu/Makefile
===================================================================
RCS file: /usr2/ncvs/src/gnu/libexec/uucp/cu/Makefile,v
retrieving revision 1.8
retrieving revision 1.9
diff -u -r1.8 -r1.9
--- gnu/libexec/uucp/cu/Makefile	1999/08/27 23:33:06	1.8
+++ gnu/libexec/uucp/cu/Makefile	2001/09/09 04:54:09	1.9
@@ -12,6 +12,7 @@
 DPADD+=		$(LIBUNIX) $(LIBUUCONF) $(LIBUUCP)
 CFLAGS+=	-I$(.CURDIR)/../common_sources\
 		-DVERSION=\"$(VERSION)\"
+INSTALLFLAGS+= -fschg
 
 .include <bsd.prog.mk>
 .PATH: $(.CURDIR)/../common_sources
Index: gnu/libexec/uucp/uucp/Makefile
===================================================================
RCS file: /usr2/ncvs/src/gnu/libexec/uucp/uucp/Attic/Makefile,v
retrieving revision 1.6
retrieving revision 1.7
diff -u -r1.6 -r1.7
--- gnu/libexec/uucp/uucp/Makefile	1999/08/27 23:33:55	1.6
+++ gnu/libexec/uucp/uucp/Makefile	2001/09/09 04:54:10	1.7
@@ -11,6 +11,7 @@
 DPADD+=		$(LIBUNIX) $(LIBUUCONF) $(LIBUUCP)
 CFLAGS+=	-I$(.CURDIR)/../common_sources\
 		-DVERSION=\"$(VERSION)\"
+INSTALLFLAGS+= -fschg
 
 .include <bsd.prog.mk>
 .PATH: $(.CURDIR)/../common_sources
Index: gnu/libexec/uucp/uuname/Makefile
===================================================================
RCS file: /usr2/ncvs/src/gnu/libexec/uucp/uuname/Attic/Makefile,v
retrieving revision 1.5
retrieving revision 1.6
diff -u -r1.5 -r1.6
--- gnu/libexec/uucp/uuname/Makefile	1999/08/27 23:33:58	1.5
+++ gnu/libexec/uucp/uuname/Makefile	2001/09/09 04:54:10	1.6
@@ -11,7 +11,7 @@
 DPADD+=		$(LIBUNIX) $(LIBUUCONF) $(LIBUUCP)
 CFLAGS+=	-I$(.CURDIR)/../common_sources\
 		-DVERSION=\"$(VERSION)\"
-
+INSTALLFLAGS+= -fschg
 
 .include <bsd.prog.mk>
 .PATH: $(.CURDIR)/../common_sources
Index: gnu/libexec/uucp/uustat/Makefile
===================================================================
RCS file: /usr2/ncvs/src/gnu/libexec/uucp/uustat/Attic/Makefile,v
retrieving revision 1.5
retrieving revision 1.6
diff -u -r1.5 -r1.6
--- gnu/libexec/uucp/uustat/Makefile	1999/08/27 23:34:02	1.5
+++ gnu/libexec/uucp/uustat/Makefile	2001/09/09 04:54:10	1.6
@@ -13,6 +13,7 @@
 CFLAGS+=	-I$(.CURDIR)/../common_sources\
 		-DOWNER=\"$(owner)\"\
 		-DVERSION=\"$(VERSION)\"
+INSTALLFLAGS+= -fschg
 
 .include <bsd.prog.mk>
 .PATH: $(.CURDIR)/../common_sources
Index: gnu/libexec/uucp/uux/Makefile
===================================================================
RCS file: /usr2/ncvs/src/gnu/libexec/uucp/uux/Attic/Makefile,v
retrieving revision 1.6
retrieving revision 1.7
diff -u -r1.6 -r1.7
--- gnu/libexec/uucp/uux/Makefile	1999/08/27 23:34:05	1.6
+++ gnu/libexec/uucp/uux/Makefile	2001/09/09 04:54:10	1.7
@@ -11,6 +11,7 @@
 DPADD+=		$(LIBUNIX) $(LIBUUCONF) $(LIBUUCP)
 CFLAGS+=	-I$(.CURDIR)/../common_sources\
 		-DVERSION=\"$(VERSION)\"
+INSTALLFLAGS+= -fschg
 
 .include <bsd.prog.mk>
 .PATH: $(.CURDIR)/../common_sources
Index: usr.bin/tip/tip/Makefile
===================================================================
RCS file: /usr2/ncvs/src/usr.bin/tip/tip/Makefile,v
retrieving revision 1.11
retrieving revision 1.12
diff -u -r1.11 -r1.12
--- usr.bin/tip/tip/Makefile	2001/03/27 10:52:16	1.11
+++ usr.bin/tip/tip/Makefile	2001/09/09 04:54:09	1.12
@@ -21,6 +21,7 @@
 MAN=	tip.1 modems.5
 SRCS=	acu.c acutab.c cmds.c cmdtab.c cu.c hunt.c log.c partab.c \
 	remote.c tip.c tipout.c value.c vars.c
+INSTALLFLAGS+= -fschg
 
 BINDIR?=	/usr/bin
 BINOWN=		uucp
Index: etc/periodic/daily/410.status-uucp
===================================================================
RCS file: /usr2/ncvs/src/etc/periodic/daily/Attic/410.status-uucp,v
retrieving revision 1.5
retrieving revision 1.6
diff -u -r1.5 -r1.6
--- etc/periodic/daily/410.status-uucp	2000/09/14 17:19:10	1.5
+++ etc/periodic/daily/410.status-uucp	2001/09/09 05:53:01	1.6
@@ -27,7 +27,7 @@
 	    echo ""
 	    echo "UUCP status:"
 
-	    uustat -a && rc=0 || rc=3
+	    (echo "/usr/bin/uustat -a" | su -fm uucp ) && rc=0 || rc=3
 	fi;;
 
     *)  rc=0;;
