Index: sys/kern/kern_ktrace.c
===================================================================
RCS file: /home/ncvs/src/sys/kern/kern_ktrace.c,v
retrieving revision 1.35.2.5
retrieving revision 1.35.2.6
diff -u -r1.35.2.5 -r1.35.2.6
--- sys/kern/kern_ktrace.c	24 Oct 2001 19:14:51 -0000	1.35.2.5
+++ sys/kern/kern_ktrace.c	5 Jul 2002 22:36:38 -0000	1.35.2.6
@@ -571,7 +571,8 @@
 	     target->p_ruid == target->p_svuid &&
 	     caller->p_rgid == target->p_rgid &&	/* XXX */
 	     target->p_rgid == target->p_svgid &&
-	     (targetp->p_traceflag & KTRFAC_ROOT) == 0) ||
+	     (targetp->p_traceflag & KTRFAC_ROOT) == 0 &&
+	     (targetp->p_flag & P_SUGID) == 0) ||
 	     caller->pc_ucred->cr_uid == 0)
 		return (1);
 
