Index: findterm.c
===================================================================
RCS file: /usr/cvs/src/lib/libmytinfo/Attic/findterm.c,v
retrieving revision 1.3
diff -u -r1.3 findterm.c
--- findterm.c	1997/08/13 01:21:36	1.3
+++ findterm.c	2000/04/25 16:58:19
@@ -242,7 +242,7 @@
 			} else {
 				s = path->file;
 				d = buf;
-				while(*s != '\0' && *s != ':')
+				while(*s != '\0' && *s != ':' && d - buf < MAX_LINE - 1)
 					*d++ = *s++;
 				*d = '\0';
 				if (_tmatch(buf, name)) {
@@ -259,7 +259,7 @@
 			} else {
 				s = path->file;
 				d = buf;
-				while(*s != '\0' && *s != ',')
+				while(*s != '\0' && *s != ',' && d - buf < MAX_LINE - 1)
 					*d++ = *s++;
 				*d = '\0';
 				if (_tmatch(buf, name)) {
