Index: sys/net/if.c
===================================================================
RCS file: /home/ncvs/src/sys/net/if.c,v
retrieving revision 1.85.2.28
diff -u -p -r1.85.2.28 if.c
--- sys/net/if.c	12 Dec 2004 20:12:50 -0000	1.85.2.28
+++ sys/net/if.c	14 Apr 2005 22:10:46 -0000
@@ -1326,6 +1326,12 @@ ifconf(cmd, data)
 		char workbuf[64];
 		int ifnlen, addrs;
 
+		/*
+		 * Zero the ifr_name buffer to make sure we don't
+		 * disclose the contents of the stack.
+		 */
+		memset(ifr.ifr_name, 0, sizeof(ifr.ifr_name));
+
 		if (space <= sizeof (ifr))
 			break;
 		ifnlen = snprintf(workbuf, sizeof(workbuf),
