Index: contrib/bind9/lib/dns/rdata/generic/keydata_65533.c
===================================================================
--- contrib/bind9/lib/dns/rdata/generic/keydata_65533.c	(revision 253461)
+++ contrib/bind9/lib/dns/rdata/generic/keydata_65533.c	(working copy)
@@ -176,7 +176,7 @@
 	UNUSED(options);
 
 	isc_buffer_activeregion(source, &sr);
-	if (sr.length < 4)
+	if (sr.length < 16)
 		return (ISC_R_UNEXPECTEDEND);
 
 	isc_buffer_forward(source, sr.length);
