Index: sys/kern/sysv_shm.c
===================================================================
RCS file: /home/ncvs/src/sys/kern/sysv_shm.c,v
retrieving revision 1.89
diff -C3 -c -r1.89 sysv_shm.c
*** sys/kern/sysv_shm.c	7 Nov 2003 04:47:14 -0000	1.89
--- sys/kern/sysv_shm.c	3 Feb 2004 15:12:26 -0000
***************
*** 378,380 ****
--- 378,381 ----
  	rv = vm_map_find(&p->p_vmspace->vm_map, shm_handle->shm_object,
  		0, &attach_va, size, (flags & MAP_FIXED)?0:1, prot, prot, 0);
  	if (rv != KERN_SUCCESS) {
+ 		vm_object_deallocate(shm_handle->shm_object);
