--- sys/dev/nvmf/controller/nvmft_controller.c
+++ sys/dev/nvmf/controller/nvmft_controller.c
@@ -227,7 +227,7 @@
 	if (ctrlr == NULL) {
 		mtx_unlock(&np->lock);
 		printf("NVMFT: Nonexistent controller %u for I/O queue %u from %.*s\n",
-		    ctrlr->cntlid, qid, (int)sizeof(data->hostnqn),
+		    cntlid, qid, (int)sizeof(data->hostnqn),
 		    data->hostnqn);
 		nvmft_connect_invalid_parameters(qp, cmd, true,
 		    offsetof(struct nvmf_fabric_connect_data, cntlid));

