--- lib/libcasper/services/cap_net/cap_net.c.orig
+++ lib/libcasper/services/cap_net/cap_net.c
@@ -1232,7 +1232,7 @@
 	oldfunclimits = NULL;
 	if (oldlimits != NULL) {
 		oldfunclimits = dnvlist_get_nvlist(oldlimits,
-		    LIMIT_NV_ADDR2NAME, NULL);
+		    LIMIT_NV_NAME2ADDR, NULL);
 	}
 
 	cookie = NULL;
--- lib/libcasper/services/cap_net/tests/net_test.c.orig
+++ lib/libcasper/services/cap_net/tests/net_test.c
@@ -783,6 +783,18 @@
 	limit = cap_net_limit_init(capnet, CAPNET_NAME2ADDR);
 	ATF_REQUIRE(cap_net_limit(limit) != 0);
 
+	/* Try to extend the limit. */
+	limit = cap_net_limit_init(capnet, CAPNET_NAME2ADDR);
+	ATF_REQUIRE(limit != NULL);
+	cap_net_limit_name2addr(limit, TEST_DOMAIN_1, NULL);
+	ATF_REQUIRE(cap_net_limit(limit) != 0);
+
+	limit = cap_net_limit_init(capnet, CAPNET_NAME2ADDR);
+	ATF_REQUIRE(limit != NULL);
+	cap_net_limit_name2addr(limit, TEST_DOMAIN_0, NULL);
+	cap_net_limit_name2addr(limit, TEST_DOMAIN_1, NULL);
+	ATF_REQUIRE(cap_net_limit(limit) != 0);
+
 	cap_close(capnet);
 }
 
