--- usr.sbin/pw/pwupd.c	2001/08/20 15:09:34
+++ usr.sbin/pw/pwupd.c	2001/12/20 16:03:04
@@ -176,7 +176,7 @@
 			 */
 			if (pwd != NULL)
 				fmtpwentry(pwbuf, pwd, PWF_MASTER);
-			rc = fileupdate(getpwpath(_MASTERPASSWD), 0644, pwbuf, pfx, l, mode);
+			rc = fileupdate(getpwpath(_MASTERPASSWD), 0600, pwbuf, pfx, l, mode);
 			if (rc == 0) {
 #ifdef HAVE_PWDB_U
 				if (mode == UPD_DELETE || isrename)
