Index: sys/ufs/ffs/ffs_inode.c
===================================================================
RCS file: /home/ncvs/src/sys/ufs/ffs/ffs_inode.c,v
retrieving revision 1.93.2.1
diff -u -r1.93.2.1 ffs_inode.c
--- sys/ufs/ffs/ffs_inode.c	31 Jan 2005 23:26:59 -0000	1.93.2.1
+++ sys/ufs/ffs/ffs_inode.c	11 Mar 2005 14:24:46 -0000
@@ -274,6 +274,7 @@
 			    IO_EXT | IO_NORMAL : IO_NORMAL);
 			vinvalbuf(ovp, needextclean ? 0 : V_NORMAL,
 			    cred, td, 0, 0);
+			vnode_pager_setsize(vp, 0);
 			oip->i_flag |= IN_CHANGE | IN_UPDATE;
 			return (ffs_update(ovp, 0));
 		}
