Index: contrib/cvs/src/server.c
===================================================================
RCS file: /home/ncvs/src/contrib/cvs/src/server.c,v
retrieving revision 1.13.2.2
diff -c -r1.13.2.2 server.c
*** server.c	28 Oct 2001 21:32:10 -0000	1.13.2.2
--- server.c	4 Feb 2003 13:16:15 -0000
***************
*** 984,992 ****
  	return;
      }
  
-     if (dir_name != NULL)
- 	free (dir_name);
- 
      dir_len = strlen (dir);
  
      /* Check for a trailing '/'.  This is not ISDIRSEP because \ in the
--- 984,989 ----
***************
*** 1002,1007 ****
--- 999,1007 ----
  	return;
      }
  
+     if (dir_name != NULL)
+ 	free (dir_name);
+ 
      dir_name = malloc (strlen (server_temp_dir) + dir_len + 40);
      if (dir_name == NULL)
      {
***************
*** 4738,4745 ****
    REQ_LINE("Max-dotdot", serve_max_dotdot, 0),
    REQ_LINE("Static-directory", serve_static_directory, 0),
    REQ_LINE("Sticky", serve_sticky, 0),
!   REQ_LINE("Checkin-prog", serve_checkin_prog, 0),
!   REQ_LINE("Update-prog", serve_update_prog, 0),
    REQ_LINE("Entry", serve_entry, RQ_ESSENTIAL),
    REQ_LINE("Kopt", serve_kopt, 0),
    REQ_LINE("Checkin-time", serve_checkin_time, 0),
--- 4738,4745 ----
    REQ_LINE("Max-dotdot", serve_max_dotdot, 0),
    REQ_LINE("Static-directory", serve_static_directory, 0),
    REQ_LINE("Sticky", serve_sticky, 0),
!   REQ_LINE("Checkin-prog", serve_noop, 0),
!   REQ_LINE("Update-prog", serve_noop, 0),
    REQ_LINE("Entry", serve_entry, RQ_ESSENTIAL),
    REQ_LINE("Kopt", serve_kopt, 0),
    REQ_LINE("Checkin-time", serve_checkin_time, 0),
