Index: usr.sbin/freebsd-update/freebsd-update.sh
===================================================================
--- usr.sbin/freebsd-update/freebsd-update.sh	(revision 257878)
+++ usr.sbin/freebsd-update/freebsd-update.sh	(revision 257879)
@@ -2884,10 +2884,10 @@
 	install_delete INDEX-OLD INDEX-NEW || return 1
 
 	# Remove old directories
+	grep -vE '^/boot/' $1/INDEX-NEW |
+	    grep -E '^[^|]+\|d\|' > INDEX-NEW
 	grep -vE '^/boot/' $1/INDEX-OLD |
 	    grep -E '^[^|]+\|d\|' > INDEX-OLD
-	grep -vE '^/boot/' $1/INDEX-OLD |
-	    grep -E '^[^|]+\|d\|' > INDEX-OLD
 	install_delete INDEX-OLD INDEX-NEW || return 1
 
 	# Remove temporary files
