--- release/packages/generate-set-ucl.lua.orig
+++ release/packages/generate-set-ucl.lua
@@ -16,7 +16,7 @@
 local ucl = require("ucl")
 
 -- This parser is the output UCL we want to build.
-local parser = ucl.parser()
+local parser = ucl.parser(0)
 
 if #arg < 1 then
 	io.stderr:write(arg[0] .. ": missing template filename\n")
--- release/packages/generate-ucl.lua.orig
+++ release/packages/generate-ucl.lua
@@ -124,7 +124,7 @@
 local pkgversion = nil
 
 -- This parser is the output UCL we want to build.
-local parser = ucl.parser()
+local parser = ucl.parser(0)
 
 -- Set any $VARIABLES from the command line in the parser.  This causes ucl to
 -- automatically replace them when we load the source ucl.
