--- el.c	1999/08/20 01:17:12	1.6
+++ el.c	2000/05/22 06:01:31	1.8
@@ -290,13 +294,10 @@
     char *ptr, path[MAXPATHLEN];
 
     if (fname == NULL) {
-	fname = &elpath[1];
-	if ((fp = fopen(fname, "r")) == NULL) {
-	    if (issetugid() != 0 || (ptr = getenv("HOME")) == NULL)
-		return -1;
-	    (void)snprintf(path, sizeof(path), "%s%s", ptr, elpath);
-	    fname = path;
-	}
+	if (issetugid() != 0 || (ptr = getenv("HOME")) == NULL)
+	    return -1;
+	(void) snprintf(path, sizeof(path), "%s%s", ptr, elpath);
+	fname = path;
     }
 
     if ((fp = fopen(fname, "r")) == NULL)
