Index: pw_util.c
===================================================================
RCS file: /usr/home/ncvs/src/usr.sbin/vipw/pw_util.c,v
retrieving revision 1.17
retrieving revision 1.18
diff -u -r1.17 -r1.18
--- pw_util.c	1999/08/28 01:20:31	1.17
+++ pw_util.c	2000/07/12 00:49:40	1.18
@@ -250,7 +250,7 @@
 	extern int _use_yp;
 #endif /* YP */
 	if (err)
-		warn(name);
+		warn("%s", name);
 #ifdef YP
 	if (_use_yp)
 		warnx("NIS information unchanged");
