Index: sys/netinet6/in6.c
===================================================================
RCS file: /home/ncvs/src/sys/netinet6/in6.c,v
retrieving revision 1.109
diff -p -u -I__FBSDID -I$FreeBSD -u -r1.109 in6.c
--- sys/netinet6/in6.c	27 May 2009 14:11:23 -0000	1.109
+++ sys/netinet6/in6.c	8 Jun 2009 18:02:59 -0000
@@ -215,6 +215,7 @@ in6_control(struct socket *so, u_long cm
 	case SIOCSRTRFLUSH_IN6:
 	case SIOCSDEFIFACE_IN6:
 	case SIOCSIFINFO_FLAGS:
+	case SIOCSIFINFO_IN6:
 		if (td != NULL) {
 			error = priv_check(td, PRIV_NETINET_ND6);
 			if (error)
@@ -223,7 +224,6 @@ in6_control(struct socket *so, u_long cm
 		/* FALLTHROUGH */
 	case OSIOCGIFINFO_IN6:
 	case SIOCGIFINFO_IN6:
-	case SIOCSIFINFO_IN6:
 	case SIOCGDRLST_IN6:
 	case SIOCGPRLST_IN6:
 	case SIOCGNBRINFO_IN6:
