Index: contrib/bind9/lib/dns/opensslrsa_link.c
===================================================================
--- contrib/bind9/lib/dns/opensslrsa_link.c	(revision 187056)
+++ contrib/bind9/lib/dns/opensslrsa_link.c	(working copy)
@@ -246,7 +246,7 @@
 
 	status = RSA_verify(type, digest, digestlen, sig->base,
 			    RSA_size(rsa), rsa);
-	if (status == 0)
+	if (status != 1)
 		return (dst__openssl_toresult(DST_R_VERIFYFAILURE));
 
 	return (ISC_R_SUCCESS);
Index: contrib/bind9/lib/dns/openssldsa_link.c
===================================================================
--- contrib/bind9/lib/dns/openssldsa_link.c	(revision 187056)
+++ contrib/bind9/lib/dns/openssldsa_link.c	(working copy)
@@ -133,7 +133,7 @@
 
 	status = DSA_do_verify(digest, ISC_SHA1_DIGESTLENGTH, dsasig, dsa);
 	DSA_SIG_free(dsasig);
-	if (status == 0)
+	if (status != 1)
 		return (dst__openssl_toresult(DST_R_VERIFYFAILURE));
 
 	return (ISC_R_SUCCESS);
