Index: etc/rc.d/jail
===================================================================
RCS file: /home/ncvs/src/etc/rc.d/jail,v
retrieving revision 1.23.2.7.2.1
diff -u -d -r1.23.2.7.2.1 jail
--- etc/rc.d/jail	11 Jan 2007 18:17:24 -0000	1.23.2.7.2.1
+++ etc/rc.d/jail	27 Feb 2007 20:47:59 -0000
@@ -331,8 +331,8 @@
 				echo ${_jail_id} > /var/run/jail_${_jail}.id
 			else
 				jail_umount_fs
-				if [ -n "${jail_interface}" ]; then
-					ifconfig ${jail_interface} -alias ${jail_ip}
+				if [ -n "${_interface}" ]; then
+					ifconfig ${_interface} -alias ${_ip}
 				fi
 				echo " cannot start jail \"${_jail}\": "
 				tail +2 ${_tmp_jail}
