Index: lib/libz/inftrees.h
===================================================================
RCS file: /home/ncvs/src/lib/libz/inftrees.h,v
retrieving revision 1.1.1.5
diff -u -r1.1.1.5 inftrees.h
--- lib/libz/inftrees.h	30 Jun 2004 23:43:27 -0000	1.1.1.5
+++ lib/libz/inftrees.h	23 Jul 2005 13:08:30 -0000
@@ -36,12 +36,12 @@
  */
 
 /* Maximum size of dynamic tree.  The maximum found in a long but non-
-   exhaustive search was 1004 code structures (850 for length/literals
-   and 154 for distances, the latter actually the result of an
+   exhaustive search was 1444 code structures (852 for length/literals
+   and 592 for distances, the latter actually the result of an
    exhaustive search).  The true maximum is not known, but the value
    below is more than safe. */
-#define ENOUGH 1440
-#define MAXD 154
+#define ENOUGH 2048
+#define MAXD 592
 
 /* Type of code to build for inftable() */
 typedef enum {
