Index: periodic.sh
===================================================================
RCS file: /home/ncvs/src/usr.sbin/periodic/periodic.sh,v
retrieving revision 1.9.2.5
diff -u -r1.9.2.5 periodic.sh
--- periodic.sh	2000/09/20 20:17:21	1.9.2.5
+++ periodic.sh	2000/12/20 14:44:22
@@ -27,7 +27,7 @@
 
 host=`hostname`
 export host
-tmp_output=${TMPDIR:-/tmp}/periodic.$$
+tmp_output=`mktemp ${TMPDIR:-/tmp}/periodic.XXXXXXXXXX`
 
 # Execute each executable file in the directory list.  If the x bit is not
 # set, assume the user didn't really want us to muck with it (it's a
@@ -89,7 +89,7 @@
                       esac
                       [ $output = TRUE ] && { cat $tmp_output; empty=FALSE; }
                     fi
-                    rm -f $tmp_output
+		    cp /dev/null $tmp_output
                 fi
             done
         done
@@ -100,3 +100,4 @@
         fi
     } | eval $pipe
 done
+rm -f $tmp_output
