Index: crypto/openssh/channels.c
===================================================================
RCS file: /home/ncvs/src/crypto/openssh/channels.c,v
retrieving revision 1.23
diff -u -r1.23 channels.c
--- crypto/openssh/channels.c	30 Sep 2006 13:38:04 -0000	1.23
+++ crypto/openssh/channels.c	15 Apr 2008 19:09:48 -0000
@@ -2895,9 +2895,6 @@
 				debug2("bind port %d: %.100s", port, strerror(errno));
 				close(sock);
 
-				if (ai->ai_next)
-					continue;
-
 				for (n = 0; n < num_socks; n++) {
 					close(socks[n]);
 				}
